VC/Zoom: Add meeting registration sync

moliholy opened the pull request:
This PR enables registration in zoom meetings when they get registered in indico events with an associated zoom meeting.

moliholy committed:
feat: enable registrations on zoom when users get registered in indico

moliholy committed:
Fix wrong function usage

moliholy committed:
Fix ruff issues

moliholy committed:
feat: enable auto-registration

moliholy committed:
fix: use fine-grained exception handling

moliholy committed:
feat: add system-wide setting to enable registrations

moliholy committed:
test: add tests for zoom meeting registration

moliholy committed:
fix: more ruff issues

moliholy committed:
Update the list of scopes

moliholy commented:
@ThiefMaster friendly ping 🙏

moliholy committed:
fix: correct logger indentation in registration sync

moliholy committed:
feat: add scope validation for auto-registration setting

moliholy committed:
fix: add passcode visibility recommendation to auto-registration

moliholy committed:
fix: use enterprise email for zoom registration

moliholy committed:
fix: make remove keyword-only in _sync_registration

moliholy committed:
feat: handle registration state changes and form deletion

moliholy committed:
refactor: batch registration sync via flask.g and after_process

moliholy committed:
feat: sync existing registrations on vc room creation

moliholy committed:
fix: ruff and isort issues

moliholy commented on vc_zoom/indico_vc_zoom/plugin.py:
```diff
@@ -91,6 +91,11 @@ class PluginSettingsForm(VCPluginSettingsFormBase):
+    auto_register = BooleanField(_('Automatic registration'),
+                                 widget=SwitchWidget(),
+                                 description=_('Automatically register Indico registrants in Zoom meetings/webinars '
+                                               'once their Indico registration is complete.'))
```
Done in [4bb83393](https://github.com/indico/indico-plugins/pull/292/commits/4bb83393).

moliholy commented on vc_zoom/indico_vc_zoom/plugin.py:
```diff
@@ -621,3 +630,76 @@ def _event_metadata_postprocess(self, sender, event, data, user=None, skip_acces
+
+            if should_be_registered:
+                data = {
+                    'email': registration.email,
```
Done in [d58b4ff2](https://github.com/indico/indico-plugins/pull/292/commits/d58b4ff2).

moliholy committed:
feat: use batch API for adding multiple registrants

moliholy committed:
fix: skip zoom removal if user has another active registration

ThiefMaster committed:
Cleanup tests

- indentation
- usefixtures instead of unused arg

moliholy committed:
fix: improve Zoom token scope handling

moliholy committed:
feat: add plugin and room controls for auto registration

moliholy commented on vc_zoom/indico_vc_zoom/api/client.py:
```diff
@@ -242,13 +310,54 @@ def update_webinar(self, webinar_id, data):
+    try:
+        access_token, *_rest = get_zoom_token(config, for_config_check=True)
+    except Exception:
+        return set()
```
I guess only a network error. Fixed in [e1117a2](https://github.com/indico/indico-plugins/pull/292/commits/e1117a26f7468ef9a5c2b2a8bce9a0aef71a4a7d).

moliholy commented on vc_zoom/indico_vc_zoom/api/client.py:
```diff
@@ -242,13 +310,54 @@ def update_webinar(self, webinar_id, data):
+        return set()
+
+    # get_zoom_token populates the token_cache as a side effect;
+    # read it back to access the scope field, which is not in the return tuple.
```
Was thinking about that, actually. Done in [e1117a2](https://github.com/indico/indico-plugins/pull/292/commits/e1117a26f7468ef9a5c2b2a8bce9a0aef71a4a7d).

moliholy commented on vc_zoom/README.md:
```diff
@@ -188,12 +188,16 @@ The scopes to select when creating the app are:
 - `meeting:update:meeting:admin`
 - `meeting:delete:meeting:admin`
+- `meeting:read:list_registrants:admin` (required for automatic registration feature)
+- `meeting:write:registrant:admin` (required for automatic registration feature)
```
Done in [e1117a2](https://github.com/indico/indico-plugins/pull/292/commits/e1117a26f7468ef9a5c2b2a8bce9a0aef71a4a7d)

moliholy commented on vc_zoom/indico_vc_zoom/plugin.py:
```diff
@@ -129,18 +170,37 @@ def validate_authenticators(self, field):
-                                      for_config_check=True)
-        if ok:
+        access_token, url, msg = get_zoom_token(config, for_config_check=True)
+        if access_token:
```
Done in [e1117a2](https://github.com/indico/indico-plugins/pull/292/commits/e1117a26f7468ef9a5c2b2a8bce9a0aef71a4a7d).

moliholy commented on vc_zoom/indico_vc_zoom/api/client.py:
```diff
@@ -242,13 +310,54 @@ def update_webinar(self, webinar_id, data):
+        return set()
+
+    try:
+        access_token, *_rest = get_zoom_token(config, for_config_check=True)
```
Yep, indeed. Done in [e1117a2](https://github.com/indico/indico-plugins/pull/292/commits/e1117a26f7468ef9a5c2b2a8bce9a0aef71a4a7d).

moliholy commented on vc_zoom/indico_vc_zoom/plugin.py:
```diff
@@ -91,6 +122,16 @@ class PluginSettingsForm(VCPluginSettingsFormBase):
+                                               'once their Indico registration is complete. Requires the Zoom API '
+                                               'scopes {}. If webinars are enabled, add {} as well. Consider setting '
+                                               '"Passcode visibility" to "No one" so participants '
+                                               'only receive their personalized join link.')
```
OK, I see. This a a bigger one. Done in [a52803d](https://github.com/indico/indico-plugins/pull/292/commits/a52803d3583bad413ac45ea7f3c00d224b8c940e). I set it as enabled by default (if visible), and I'd recommend setting by default the passcode visibility to "No one" by default in this case too (not done, just a suggestion).

ThiefMaster replied:
my preference would be to default to off, also because many instances will have mainly meetings where "registrants" are used merely to track participants (but no open registration), and there the feature would not be ideal

moliholy replied:
Done in [ab80d76f](https://github.com/indico/indico-plugins/pull/292/commits/ab80d76f15097cc6b01936e22272eca1be00b195).

moliholy committed:
fix: satisfy lint in registration tests

moliholy committed:
feat: personalized join URLs and clone guard for auto-register

ThiefMaster commented on vc_zoom/indico_vc_zoom/plugin.py:
```diff
@@ -129,18 +168,37 @@ def validate_authenticators(self, field):
+        if not all(config.values()):
+            return
+        scopes = get_zoom_scopes(config)
+        if not scopes:
```
```suggestion
        if not (scopes := get_zoom_scopes(config)):
```

moliholy replied:
Done in [ab80d76f](https://github.com/indico/indico-plugins/pull/292/commits/ab80d76f15097cc6b01936e22272eca1be00b195).

ThiefMaster commented on vc_zoom/indico_vc_zoom/plugin.py:
```diff
@@ -129,18 +168,37 @@ def validate_authenticators(self, field):
+            return
+        missing_scopes = _get_missing_auto_registration_scopes(scopes, allow_webinars=self.allow_webinars.data)
+        if missing_scopes:
+            raise ValidationError(
```
```suggestion
        if missing_scopes := _get_missing_auto_registration_scopes(scopes, allow_webinars=self.allow_webinars.data)
```

moliholy replied:
Done in [ab80d76f](https://github.com/indico/indico-plugins/pull/292/commits/ab80d76f15097cc6b01936e22272eca1be00b195).

ThiefMaster commented on vc_zoom/indico_vc_zoom/plugin.py:
```diff
@@ -177,12 +236,21 @@ class ZoomPlugin(VCPluginMixin, IndicoPlugin):
     }
+    # Limited to 30 by Zoom API.
+    # See https://developers.zoom.us/docs/api/meetings/#tag/invitation--registration/post/meetings/{meetingId}/registrants
+    _BATCH_REGISTRANTS_MAX = 30
```
I would set this on the module level together with the other (scope-related) constants and not inside the plugin class. Or alternatively on the zoom client class where it would IMHO fit even better.

moliholy replied:
Done in [ab80d76f](https://github.com/indico/indico-plugins/pull/292/commits/ab80d76f15097cc6b01936e22272eca1be00b195).

ThiefMaster commented on vc_zoom/indico_vc_zoom/plugin.py:
```diff
@@ -665,3 +737,234 @@ def _event_metadata_postprocess(self, sender, event, data, user=None, skip_acces
+            return None
+
+        cache_key = None
+        if has_request_context():
```
Is this ever called outside a request context? (maybe when getting the zoom link for emails or ical?)

moliholy replied:
Yes, it's now also called from the iCal export path.

ThiefMaster commented on vc_zoom/indico_vc_zoom/plugin.py:
```diff
@@ -665,3 +737,234 @@ def _event_metadata_postprocess(self, sender, event, data, user=None, skip_acces
+
+        room_ops = self._collect_room_ops(pending)
+        if not room_ops:
+            return
```
```suggestion
        if not (pending := g.pop('zoom_pending_registrations', None)):
            return

        if not (room_ops := self._collect_room_ops(pending)):
            return
```

moliholy replied:
Done in [ab80d76f](https://github.com/indico/indico-plugins/pull/292/commits/ab80d76f15097cc6b01936e22272eca1be00b195).

ThiefMaster commented on vc_zoom/indico_vc_zoom/plugin.py:
```diff
@@ -665,3 +737,234 @@ def _event_metadata_postprocess(self, sender, event, data, user=None, skip_acces
+        from indico.modules.events.registration.models.registrations import Registration
+
+        event_ids = {assoc.event_id for assoc in vc_room.events}
+        if not event_ids:
```
```suggestion
        if not (event_ids := {assoc.event_id for assoc in vc_room.events}):
```

moliholy replied:
Done in [ab80d76f](https://github.com/indico/indico-plugins/pull/292/commits/ab80d76f15097cc6b01936e22272eca1be00b195).

ThiefMaster commented on vc_zoom/indico_vc_zoom/plugin.py:
```diff
@@ -665,3 +737,234 @@ def _event_metadata_postprocess(self, sender, event, data, user=None, skip_acces
+    def _get_registrant_email(self, registration):
+        if registration.user:
+            enterprise_email = find_enterprise_email(registration.user)
+            if enterprise_email:
```
```suggestion
            if enterprise_email := find_enterprise_email(registration.user):
```

moliholy replied:
Done in [ab80d76f](https://github.com/indico/indico-plugins/pull/292/commits/ab80d76f15097cc6b01936e22272eca1be00b195).

ThiefMaster commented on vc_zoom/indico_vc_zoom/plugin.py:
```diff
@@ -665,3 +737,234 @@ def _event_metadata_postprocess(self, sender, event, data, user=None, skip_acces
+        if registration.user:
+            enterprise_email = find_enterprise_email(registration.user)
+            if enterprise_email:
+                return enterprise_email
```
```suggestion
                return enterprise_email.lower()
```

and remove the `email = email.lower()` in all the other places where it's no longer needed.

because on Registration, User, etc. we already enforce lowercase emails in the DB via a check constraint. so the only place where we might end up with mixed-case emails is data coming from zoom (technically it should also not happen there in case of registrations because these come from indico, but for zoom user emails there's no such control)

moliholy replied:
Done in [ab80d76f](https://github.com/indico/indico-plugins/pull/292/commits/ab80d76f15097cc6b01936e22272eca1be00b195).

ThiefMaster commented on vc_zoom/README.md:
Add a changelog line for this feature

moliholy replied:
Done in [ab80d76f](https://github.com/indico/indico-plugins/pull/292/commits/ab80d76f15097cc6b01936e22272eca1be00b195).

ThiefMaster commented on vc_zoom/indico_vc_zoom/plugin.py:
```diff
@@ -564,17 +634,19 @@ def get_blueprints(self):
             'mute_host_video': self.settings.get('mute_host_video'),
             'mute_participant_video': self.settings.get('mute_participant_video'),
             'waiting_room': self.settings.get('waiting_room'),
+            'auto_register': auto_register_enabled,
```
I still prefer this to be opt-in here

```suggestion
            'auto_register': False,
```

moliholy replied:
Done in [ab80d76f](https://github.com/indico/indico-plugins/pull/292/commits/ab80d76f15097cc6b01936e22272eca1be00b195).

ThiefMaster replied:
I think somehow this got reverted in 05d8a1ac9. I still don't like the idea of having this and no-one visibility enabled by default... it would be extremely intrusive e.g. on CERN Indico where most people do not need it.

moliholy replied:
Alright [5104884](https://github.com/indico/indico-plugins/pull/292/commits/51048848542a706a09dddf41f2c9f0f84278d565).

ThiefMaster commented on vc_zoom/indico_vc_zoom/plugin.py:
Should this also take personalized links into account? This is used when someone generates an ical file for an event.

moliholy replied:
Nice catch. Done in [bc9171e](https://github.com/indico/indico-plugins/pull/292/commits/bc9171e235346a31e9d950c4622cbf3006674525).

ThiefMaster commented on vc_zoom/indico_vc_zoom/plugin.py:
```diff
@@ -42,11 +43,41 @@
                                  process_alternative_hosts, update_zoom_meeting)
 
 
+AUTO_REGISTRATION_MEETING_SCOPES = ('meeting:read:list_registrants:admin', 'meeting:write:registrant:admin')
```
Some scopes are missing here. When withdrawing a registration I got this error:

```
2026-04-13 11:23:44,573  ERROR    a10439468c6a43df  6       indico.plugin.vc_zoom     Error in API call to https://cern.zoom.us/v2/meetings/xxx/registrants/status: b'{"code":4711,"message":"Invalid access token, does not contain scopes:[meeting:update:registrant_status, meeting:update:registrant_status:admin]."}'
```

So I think `meeting:update:registrant_status:admin` needs to be added to the list of required scopes.

ThiefMaster replied:
And probably `webinar:update:registrant_status:admin` as well for webinars.

ThiefMaster replied:
I tested on our tenant and adding `meeting:update:registrant_status:admin` fixed the error for meetings.

moliholy replied:
Heh, my bad: I already had that scope by mistake in the zoom app I was testing against. Fixed in [bacc034](https://github.com/indico/indico-plugins/pull/292/commits/bacc0340689c66edfcb0be5800d7be8c1335d858).

ThiefMaster commented on vc_zoom/indico_vc_zoom/plugin.py:
```diff
@@ -42,11 +43,41 @@
+def _get_missing_auto_registration_scopes(granted_scopes, *, allow_webinars):
+    missing_scopes = []
+    if not _has_required_zoom_scopes(granted_scopes, AUTO_REGISTRATION_SCOPE_OPTIONS['meeting']):
+        missing_scopes.extend(AUTO_REGISTRATION_MEETING_SCOPES)
```
Shouldn't this only add the missing scopes to the list, not all required scopes? When I tested adding the status update scopes I mentioned in my last comment, I noticed that I got this error which also includes scopes that are already present:

<img width="675" height="204" alt="Image" src="https://github.com/user-attachments/assets/f14206cf-7467-4458-8cfd-f7578016bcbd" />

moliholy replied:
You're right, that makes much more sense. Done in [42af8f2](https://github.com/indico/indico-plugins/pull/292/commits/42af8f21f93ca716ad07273e56d3470a63b7b192).

ThiefMaster commented on vc_zoom/indico_vc_zoom/plugin.py:
```diff
@@ -91,6 +122,14 @@ class PluginSettingsForm(VCPluginSettingsFormBase):
+                                                    'individual meetings/webinars. Requires the Zoom API scopes {}. '
+                                                    'If webinars are enabled, add {} as well.')
+                                      .format(_format_zoom_scopes(AUTO_REGISTRATION_MEETING_SCOPES),
+                                              _format_zoom_scopes(AUTO_REGISTRATION_WEBINAR_SCOPES)))
```
Maybe this is overkill to list here. There's the README and if you try to enable this w/o the required scope, you already get an error. A [link to the relevant section of the plugin's README](https://github.com/indico/indico-plugins/tree/master/vc_zoom#zoom-server-to-server-oauth) could be a nice alternative?

moliholy replied:
Done in ~~[ab80d76f](https://github.com/indico/indico-plugins/pull/292/commits/ab80d76f15097cc6b01936e22272eca1be00b195)~~ [97414af](https://github.com/indico/indico-plugins/pull/292/commits/97414afb30114546a1dd2252ef647baecaf89459).

ThiefMaster commented on vc_zoom/indico_vc_zoom/plugin.py:
```diff
@@ -665,3 +737,234 @@ def _event_metadata_postprocess(self, sender, event, data, user=None, skip_acces
+                 .filter(RegistrationForm.event_id.in_(event_ids),
+                         Registration.state == RegistrationState.complete,
+                         ~Registration.is_deleted,
+                         ~RegistrationForm.is_deleted))
```
I think you need to join `Registration.event` (or `RegistrationForm.event`, shouldn't matter which one) here as well and include `~Event.is_deleted` in the filters.

moliholy replied:
Yep, makes sense. Done in [ab1af66](https://github.com/indico/indico-plugins/pull/292/commits/ab1af669b056f91f120aa17657aa9b2f2126d37b).

ThiefMaster commented on vc_zoom/indico_vc_zoom/forms.py:
```diff
@@ -106,6 +106,13 @@ class VCRoomForm(VCRoomFormBase):
                                 widget=SwitchWidget(),
                                 description=_('Participants may be kept in a waiting room by the host'))
 
+    auto_register = BooleanField(_('Automatic registration'),
```
I think this option should be disabled (and forced to be enabled) if the Zoom meeting is attached to another event where it's enabled.

Alternatively, we could simply disable the option altogether when adding a zoom meeting to an event that's already attached to other events, ie only allow toggling this when creating.

I would also consider to not allow attaching zoom meeting to an event if registration is enabled. Because we get so many edge cases from this that I think it's just not worth supporting... I can't come up with a good reason why someone would want to re-use a zoom meeting with registration in multiple events tbh.

moliholy replied:
> I would also consider to not allow attaching zoom meeting to an event if registration is enabled

Yes, I agree with that. Done in [6b167dd](https://github.com/indico/indico-plugins/pull/292/commits/6b167dd47a559cff21a7fd1184bbd36044090a2f).

ThiefMaster commented on vc_zoom/indico_vc_zoom/plugin.py:
```diff
@@ -665,3 +737,234 @@ def _event_metadata_postprocess(self, sender, event, data, user=None, skip_acces
+            else:
+                client.add_meeting_registrant(zoom_id, data)
+        except HTTPError:
+            self.logger.warning('Could not add registrant %s to Zoom %s %s',
```
Logging errors is great, bit I'd also log successes (`info` level). Otherwise it'll be a pain to debug if someone claims their personal Zoom link doesn't work - but with logging it would be trivial to match the removal (or addition) to the action on Indico that triggered it.

moliholy replied:
Heh, indeed. I am usually so focused on not making mistakes and if so spotting them quickly to get PRs in proper shape that I tend to forget what a non-failing workflow looks like 😆 .

Done in [a7361cd](https://github.com/indico/indico-plugins/pull/292/commits/a7361cddb7bd0238f10fc99645ae267426f4af81).

ThiefMaster commented on vc_zoom/indico_vc_zoom/plugin.py:
```diff
@@ -665,3 +737,234 @@ def _event_metadata_postprocess(self, sender, event, data, user=None, skip_acces
+            else:
+                client.add_meeting_registrant(zoom_id, data)
+        except HTTPError:
+            self.logger.warning('Could not add registrant %s to Zoom %s %s',
```
I would interpolate the webinar/meeting string using an f-string instead of making it a log message argument. That way tools like Sentry will see it as distinct log messages which IMHO makes sense here.


```py
zoom_type = 'webinar' if is_webinar else 'meeting'
self.logger.warning(f'Could not add registrant %s to Zoom {zoom_type} %s', ...)
```

Note that you probably need to silence a ruff warning when doing this, since *usually* it's considered a bad practice. I see it more like that dynamic parts should NOT he interpolated (because you don't want tons of distinct log messages on Sentry), but parts that are quite static ("webinar/meeting") actually make sense to interpolate.

moliholy replied:
Done in [ab80d76f](https://github.com/indico/indico-plugins/pull/292/commits/ab80d76f15097cc6b01936e22272eca1be00b195).

moliholy committed:
address review feedback: walrus operators, lowercase email, constant and defaults cleanup

- Use walrus operators in validate_allow_auto_register,
  _flush_pending_registrations, _has_other_active_room_registration,
  and _get_registrant_email
- Return enterprise_email.lower() in _get_registrant_email; remove
  redundant .lower() calls where email originates from the DB
- Move BATCH_REGISTRANTS_MAX to module level alongside other scope
  constants
- Default auto_register to False on new rooms (opt-in, not opt-out)
- Simplify allow_auto_register description; refer to the README
  instead of listing scopes inline
- Use f-string for static webinar/meeting type in log messages so
  Sentry groups them as distinct events
- Add 3.3.8 changelog entry for automatic Zoom registration

moliholy committed:
fix: missing blank line and auto_register default in form defaults

- Add missing blank line before init() (E301)
- Default auto_register to follow allow_auto_register plugin setting
  instead of always False

moliholy committed:
fix: link README reference in allow_auto_register description

moliholy committed:
feat: use personalized join URL in iCal export

moliholy committed:
fix: add missing registrant status scopes

moliholy committed:
fix: report only truly missing scopes in validation error

moliholy committed:
test: update and extend scope validation tests

moliholy committed:
fix: remove unused import in scope_test

moliholy committed:
fix: filter deleted events in active room registration check

moliholy committed:
fix: block attach/clone when auto_register is enabled

moliholy committed:
feat: log successful registrant add/remove at info level

moliholy committed:
test: update clone tests to reflect auto_register blocking behavior

moliholy commented:
@ThiefMaster OK... I think that's all.

ThiefMaster commented on vc_zoom/README.md:
```diff
@@ -11,6 +11,11 @@
 
 ## Changelog
 
+### 3.3.8
```
3.3.7 has not been released yet, so it should go there. But actually I realized just now that neither has been 3.3.6 (which kind of makes sense, these small changes were not that important to justify a new plugin release).

So please move both this changelog entry and the language interpretation changelog entry to 3.3.6 and set `version = '3.3.6'` in pyproject.toml

moliholy replied:
OK, all merged into 3.3.6 in [705f141](https://github.com/indico/indico-plugins/pull/292/commits/705f141a90b356542f4390287033631b5861aad3).

ThiefMaster commented on vc_zoom/indico_vc_zoom/plugin.py:
```diff
@@ -42,11 +43,48 @@
+    for key in keys:
+        scope_options = AUTO_REGISTRATION_SCOPE_OPTIONS[key]
+        if not _has_required_zoom_scopes(granted_scopes, scope_options):
+            best_missing = min((required - granted_scopes for required in scope_options), key=len)
```
Can this result in legacy scopes being proposed?

moliholy replied:
Indeed. Addressed in [7ee10e6](https://github.com/indico/indico-plugins/pull/292/commits/7ee10e641721c1c2c2cc54aa2e56130686d7783b).

ThiefMaster commented on vc_zoom/indico_vc_zoom/plugin.py:
```diff
@@ -665,3 +754,246 @@ def _event_metadata_postprocess(self, sender, event, data, user=None, skip_acces
+    def _has_other_active_room_registration(self, vc_room, email, exclude_ids):
+        from indico.modules.events.models.events import Event
+        from indico.modules.events.registration.models.forms import RegistrationForm
+        from indico.modules.events.registration.models.registrations import Registration
```
These imports all look safe to have on the top level w/o introducing circular imports.

moliholy replied:
Fixed in [7ee10e6](https://github.com/indico/indico-plugins/pull/292/commits/7ee10e641721c1c2c2cc54aa2e56130686d7783b).

ThiefMaster commented on vc_zoom/tests/scope_test.py:
```diff
@@ -0,0 +1,38 @@
+
+def test_auto_registration_scope_check_requires_webinar_scopes_when_enabled():
+    # When no webinar scopes are present, the function reports the option set with fewest missing scopes
+    # (legacy: 2 scopes) rather than the modern set (3 scopes)
```
OK this seems to answer my question asked in another comment... I don't think it's a good idea to ever propose legacy scopes unless the app is currently already using legacy scopes.

moliholy replied:
Scopes controlled in [7ee10e6](https://github.com/indico/indico-plugins/pull/292/commits/7ee10e641721c1c2c2cc54aa2e56130686d7783b).

moliholy committed:
fix: revert to release 3.3.6

moliholy committed:
fix: do not propose legacy zoom scopes unless already in use

ThiefMaster commented on vc_zoom/indico_vc_zoom/plugin.py:
```diff
@@ -665,3 +760,239 @@ def _event_metadata_postprocess(self, sender, event, data, user=None, skip_acces
+            else:
+                client.update_meeting_registrants_status(zoom_id, status_data)
+        except HTTPError:
+            self.logger.warning(f'Could not remove registrants from Zoom {zoom_type} %s',  # noqa: G004
```
Include the list of emails here

ThiefMaster replied:
the registration id would also be useful (both here and when adding)

moliholy replied:
Done in [a5145e0](https://github.com/indico/indico-plugins/pull/292/commits/a5145e0dd626301e325080a17aeb44319bc56c76).

ThiefMaster replied:
ah sorry, I meant the indico registration id, not the zoom registration id. having the latter as well isn't necessarily bad though

moliholy replied:
Done in [9f581bf](https://github.com/indico/indico-plugins/pull/292/commits/9f581bf397ea980459b5d80bafe0aab69fb6e31e).

ThiefMaster commented on vc_zoom/indico_vc_zoom/plugin.py:
```diff
@@ -665,3 +760,239 @@ def _event_metadata_postprocess(self, sender, event, data, user=None, skip_acces
+            return
+        if not vc_room.data.get('auto_register'):
+            return
+        for regform in event.registration_forms:
+            for registration in regform.active_registrations:
+                if registration.state == RegistrationState.complete:
+                    self._queue_registration_sync(registration, remove=False)
```
This also needs to be done when enabling the auto-registration option so existing registrants get added on Zoom.

moliholy replied:
Done in [a5145e0](https://github.com/indico/indico-plugins/pull/292/commits/a5145e0dd626301e325080a17aeb44319bc56c76).

moliholy committed:
refactor: include extra data in logs

moliholy committed:
feat: log indico registration id

ThiefMaster commented on vc_zoom/indico_vc_zoom/plugin.py:
```diff
@@ -42,11 +45,52 @@
 
 
+AUTO_REGISTRATION_MEETING_SCOPES = ('meeting:read:list_registrants:admin', 'meeting:write:registrant:admin',
+                                    'meeting:update:registrant_status:admin')
```
found another missing scope: `meeting:write:batch_registrants:admin` (+ `webinar:write:batch_registrants:admin`)

moliholy replied:
Good catch. Fixed in [c1dd39f](https://github.com/indico/indico-plugins/pull/292/commits/c1dd39fde423a3cf31fa84490696f7fd9bee87ef).

moliholy committed:
fix: add missing batch_registrants scopes

ThiefMaster commented on vc_zoom/README.md:
```diff
@@ -11,12 +11,11 @@
-
 ### 3.3.6
 
+- Add automatic Zoom registration: event managers can opt-in to automatically register/unregister
```
How does this feature work for registrants that do not have an Indico account? Since Zoom does NOT send emails when you're registered, they do not have the Zoom link, and because they are not loggeg-in on Indico, it cannot show the zoom link for them in there either...

The only solutions I can think of are:

- Indico sending an email with the Zoom link when you are auto-registered
- Displaying the personal zoom links on the registration details page which you can access w/ your registration token (that's in the registration confirmation email)

moliholy replied:
Hmmm yes, that's true. I think the simplest is to just display the personal zoom link in the registration details page.

On the other hand, emails can get lost, specially if the user is registering for an event in say +6 months.

So... I guess using both should cover most of the cases. Or perhaps only displaying the zoom link would suffice 🤔 

@ThiefMaster what's your take?

ThiefMaster replied:
I'd go for the registration details page for now, and typically event organizers send emails closer to an event anyway, and those emails include the token to access the registration details anyway.

moliholy replied:
Done in [ba878db](https://github.com/indico/indico-plugins/pull/292/commits/ba878db28d6a1b2136574c8207d31ec9857d9c30).

<img width="1186" height="733" alt="Screenshot 2026-04-15 at 15 02 43" src="https://github.com/user-attachments/assets/ce9dba8e-8c41-4244-99a2-275a15f91681" />

ThiefMaster commented on vc_zoom/indico_vc_zoom/plugin.py:
```diff
@@ -318,6 +401,14 @@ def update_data_vc_room(self, vc_room, data, is_new=False):
+                for regform in event_assoc.event.registration_forms:
+                    for registration in regform.active_registrations:
+                        if registration.state == RegistrationState.complete:
+                            self._queue_registration_sync(registration, remove=False)
```
This sends all the registrations - what about getting the current list of registrations from Zoom and only sending those that are not registered yet?

moliholy replied:
Cleaner, yes. Done in [bd10f8e](https://github.com/indico/indico-plugins/pull/292/commits/bd10f8efb728530eb062914d8b1f5901e86d763b).

moliholy committed:
fix: skip existing zoom registrants on auto_register enable

ThiefMaster commented on vc_zoom/indico_vc_zoom/plugin.py:
```diff
@@ -665,3 +787,245 @@ def _event_metadata_postprocess(self, sender, event, data, user=None, skip_acces
+    def _registration_state_updated(self, registration, **kwargs):
+        if registration.state == RegistrationState.complete:
+            self._queue_registration_sync(registration, remove=False)
+        elif registration.state == RegistrationState.withdrawn:
```
```suggestion
        elif registration.state in {RegistrationState.pending, RegistrationState.withdrawn}:
```

otherwise "reset approval" doesn't remove the registration from zoom

PS: Please rebase

moliholy replied:
Done in [d892766](https://github.com/indico/indico-plugins/pull/292/commits/d8927660e6e716222a95858425b294f290dab233). Rebased as well.

moliholy committed:
fix: remove zoom registrant when registration becomes pending

moliholy committed:
feat: show personalized zoom join url on registration summary

ThiefMaster commented on vc_zoom/README.md:
```diff
@@ -15,6 +15,8 @@
 - Add automatic Zoom registration: event managers can opt-in to automatically register/unregister
   Indico participants in the corresponding Zoom meeting or webinar
+- Show the personalized Zoom join URL on the registration summary page so participants
+  without an Indico account can retrieve it from their registration token link
```
Please do not add multiple changelog entries for the same feature. I don't think this level of detail is needed here so I'd stick with the original changelog entry you added for this feature.

moliholy replied:
Yeah, too much information. Removed in [94b1247](https://github.com/indico/indico-plugins/pull/292/commits/94b1247a83f7e6ab4f973f7f11436d5f39376bbd).

moliholy committed:
fix: remove version entry

ThiefMaster commented on vc_zoom/indico_vc_zoom/templates/registration_zoom_link.html:
```diff
@@ -0,0 +1,17 @@
+            {% endif %}
+        </th>
+        <td class="regform-done-data">
+            <a href="{{ entry.url }}" target="_blank" rel="noopener noreferrer" style="word-break: break-all;">{{ entry.url }}</a>
```
I don't think this looks great. I'd rather have the usual zoom join link style...

<img width="740" height="101" alt="Image" src="https://github.com/user-attachments/assets/d42330e9-40b6-4b2c-9cab-a0df66535dbe" />

moliholy replied:
The idea was to show that the link was personalized, but yes... not ideal. Does this look better?

<img width="1097" height="703" alt="Screenshot 2026-04-15 at 16 49 05" src="https://github.com/user-attachments/assets/72b4f397-5fc3-4f90-ab72-2ac7a4b3a091" />

ThiefMaster replied:
You could add something like "(Personalized link)" behind it if you want to make it clearer...

But yes, it's better. The button like in other places would be even better IMHO. 
<img width="91" height="48" alt="image" src="https://github.com/user-attachments/assets/5d93fc0d-1b1e-4b53-9d43-98992336e95d" />

You could also try just rendering the Zoom heading and then not using two columns below but a single wider one...

moliholy replied:
Attempted to address it in [2b1298f](https://github.com/indico/indico-plugins/pull/292/commits/2b1298f84381256d2e5c1a581517ba12b7c86ccc).

<img width="1196" height="700" alt="Screenshot 2026-04-15 at 17 22 43" src="https://github.com/user-attachments/assets/a148490e-ee26-4cf2-8788-a54845fa4fcd" />


Good enough?

ThiefMaster replied:
yeah, i can live with that ;)

ThiefMaster commented on vc_zoom/indico_vc_zoom/plugin.py:
```diff
@@ -263,6 +263,7 @@ def init(self):
         self.connect(signals.vc.vc_room_created, self._vc_room_created)
         self.connect(signals.core.after_process, self._flush_pending_registrations)
         self.template_hook('event-vc-room-list-item-labels', self._render_vc_room_labels)
+        self.template_hook('before-render-registration-info', self._render_registration_zoom_link)
```
This is not ideal but it looks like we do not have a better for it. IMHO the best place would be a new template hook in `render_registration_summary` after the `receipts-list` div, but of course this would not work w/ the current Indico version - so let's maybe just try to make it look a bit nice in the place where it's being added now.

moliholy replied:
I agree. Leaving the hook for now since moving it would not be ideal right now. Happy to address it in the future.

ThiefMaster commented on vc_zoom/indico_vc_zoom/plugin.py:
```diff
@@ -862,6 +863,50 @@ def get_personalized_join_url(self, vc_room, event_vc_room, user):
             cached_urls[cache_key] = result
         return result
 
+    def _get_personalized_join_url_for_registration(self, vc_room, registration):
```
This looks quite similar to `get_personalized_join_url` - can the two be combined into a single function?!

moliholy replied:
Yep! ~~[eb8ae80](https://github.com/indico/indico-plugins/pull/292/commits/eb8ae80258886b120b3ac3f3689db398166fbe2a)~~ [a49eb46](https://github.com/indico/indico-plugins/pull/292/commits/a49eb465391d69a77be979e4f750ca5fb434c3e1).

ThiefMaster commented on vc_zoom/indico_vc_zoom/plugin.py:
```diff
@@ -862,6 +863,50 @@ def get_personalized_join_url(self, vc_room, event_vc_room, user):
+        if has_request_context():
+            cache_key = (vc_room.id, registration.id)
+            cached_urls = g.setdefault('zoom_personalized_join_urls_by_registration', {})
+            if cache_key in cached_urls:
```
Do we get any cache hits on this in the end? Because if not, then maybe the cache logic should not be there in the first place... (also in the other method)

moliholy replied:
Within a single request the same (vc_room, user) pair does get resolved more than once (join button macro + info box popup, and now also the iCal export path), so the cache is saving duplicate Zoom API round-trips on those pages.

That said, if you'd prefer keeping the code simpler I can drop the `flask.g` caching and rely on Zoo being fast enough.

ThiefMaster replied:
Yes then it makes sense to keep it. You could try simply using the `@memoize_request` on the method though - if that works it saves you some time.

moliholy replied:
Done in [6de3664](https://github.com/indico/indico-plugins/pull/292/commits/6de36641fa44d96de7bf93bf1861d769564ac3c9).

moliholy committed:
refactor: extract `_get_join_url_for_registration` helper

moliholy committed:
chore: tidy regform-done zoom join link

moliholy committed:
refactor: unify personalized join url resolution

moliholy committed:
refactor: use `memoize_request` for personalized join url

moliholy committed:
fix: default auto_register and password_visibility to off

ThiefMaster committed:
Fix changelog
